### ISSUE TITLE
[feature] Grid State - Columns Width/Position/Visibility

### ISSUE THREAD

**Issue by ghiscoding** [ghiscoding/aurelia-slickgrid#67]
**I'm submitting a feature request**

**Expected/desired behavior:**

* **What is the motivation / use case for changing the behavior?**
Extend the actual GridState Service to also provide Columns Width/Position/Visibility that user can save in Local Storage or DB which can later be re-used to recreate the grid in the exact same state as it was prior to closing the page. 

This extends the previous issue #24 (now implemented) to create a GridState Service

**Comment by ghiscoding** [comment]
SlickGrid was forked into multiple repos and I was not aware of this [GerHobblelt / SlickGrid](https://github.com/GerHobbelt/SlickGrid) and that repo does have an example with [Grid State](https://github.com/GerHobbelt/SlickGrid/blob/k0stya-rowspan/examples/example-0070-plugin-state.html) and Local Storage. That example is not fully working, the visibility (show/hide columns) is not working, however it does provide Grid State of the Column Width and Position. That sample will be of good help in achieving it in our lib.

**Comment by jmzagorski** [comment]
I know I mentioned this in another PR or issue (i'll have to look for it later to see if I talked about any details/limitations), but I accomplished this in a really old repo of mine [here](https://github.com/jmzagorski/SlickGridExtensions/blob/master/slick.ext.columnstore.js). Not sure if it can help too.

**Comment by ghiscoding** [comment]
Started working on this, I got part of it working (visibility & column order position), but it needs the entire `columnDef` to be used by `grid.setColumns(this.columnDefinitions)` which isn't great. 

Well I say isn't great, but it does have some extra bonus to use the entire column definitions, it holds the columns with all their state including anything that changed dynamically. For example, the `headerCssClass` can be changed dynamically and if we keep the column definitions in the grid state, then we could put back the grid with the last used CSS as well

A smaller approach would be to use a structure like the following:
```javascript
this.gridOptions = {
      presets: {
        columns: [
          { columnId: 'title', hidden: false, position: 0, width: 125 },
          { columnId: 'description', hidden: true, position: 2, width: 100 },
          { columnId: 'duration', hidden: false, position: 1, width: 90 },
        ],
        filters: [
          { columnId: 'duration', searchTerms: [2, 22, 44] },
          { columnId: 'usDateShort', operator: '<', searchTerms: ['4/20/25'] },
        ],
        sorters: [
          { columnId: 'duration', direction: 'DESC' },
          { columnId: 'complete', direction: 'ASC' }
        ],
      }
    };
```

OR 
I could also offer both small/big approach. Perhaps use a flag in the Grid Options to define which approach the user want to use... hmm that is interesting, I might go with that

**Comment by ghiscoding** [comment]
@jmzagorski 
I just finished working on Grid State and created a new Example 15 with Grid State, Presets and Local Storage. It is based on the new Major 2.x branch, so I can't merge until new version branch is merged.

So I decided to go with the shorter version, I don't think it's a good idea to keep the entire column in Local Storage or DB, especially if it contains Editors, Filters and other instantiated classes since we might have problem creating a JSON output for it.

So the shorter version it is, also there is no need for these 2 properties (`hidden` and `position`) since we use `grid.setColumns(presetGridColumns)` and the way it works is that if a column is not part of `presetGridColumns`, then it is considered hidden and for the position, it goes with the array index (so for example, if you want to flip 2 columns, just invert them in the preset array).

Here's the final of a valid Grid State
```javascript
this.gridOptions = {
      presets: {
        columns: [
          { columnId: 'title', width: 125 },
          { columnId: 'description', width: 100 },
          { columnId: 'duration', width: 90 },

           // not part of the presets array, column will be HIDDEN
          // you can still show it through Grid Menu or Column Picker
          // { columnId: 'duration', width: 90 },
        ],
        filters: [
          { columnId: 'duration', searchTerms: [2, 22, 44] },
          { columnId: 'usDateShort', operator: '<', searchTerms: ['4/20/25'] },
        ],
        sorters: [
          { columnId: 'duration', direction: 'DESC' },
          { columnId: 'complete', direction: 'ASC' }
        ],
      }
    };
];
```

That's it, it all works. We now have Column (Visibility, Size, Order Position, Sort & Filters)
Awesome!!!

**Comment by ghiscoding** [comment]
Implemented in my `Angular-Slickgrid` repo, you can test the [Example](https://ghiscoding.github.io/Angular-Slickgrid/#/gridstate).

It will come in `Aurelia-Slickgrid` lib **at the same time** as the new Major `2.x` release.

**Comment by ghiscoding** [comment]
closed by PR #74
